doc/design/programming_solver.md

```diff
@@ -0,0 +1,280 @@
+s.t. carpentry: sum{x in X} Carpentry[x] * ProductAmt[x] <= 80;
+s.t. demand{x in X}: ProductAmt[x] <= 40;
+```
+
+## Grammar Design
+
+In order to extend SQL to have completely same ability of AMPL, the extended syntax should be able to describe **objective and constraints** while the input data table can store the **params** for each **variable**, and the rows in the table is naturally become the **set** we defined in AMPL. 
+
+### Single Variable
+
+Then we have the below `train_table`:
+
+| product | price | materials_cost | other_cost | finishing | carpentry | max_num |
+| ------- | ----- | -------------- | ---------- | --------- | --------- | ------- |
+| soldier | 27    | 10             | 14         | 2         | 1         | 40      |
+| train   | 21    | 9              | 10         | 1         | 1         | 10000   |
+
+In the `train_table`:
+
+- The set X is row one and row two.
+- We have one variable, and the variable name strings is stored in column `product`. In cases that have multiple variables (like the example described at https://en.wikipedia.org/wiki/AMPL), the table should have multiple string columns to store the variable.
+- Other columns like `price`, `materials_cost` are all params for the corresponding variable.
+
+Then we can use below extended SQL syntax to describe above example:
+
+```sql
+SELECT * FROM train_table
+TO MAXIMIZE SUM((price - materials_cost - other_cost) * product)
+CONSTRAINT SUM(finishing * product) <= 100
+CONSTRAINT SUM(carpentry * product) <= 80
+CONSTRAINT product <= max_num
+WITH variables="product",
+     var_type="product:Integers"
+[USING glpk]
+INTO result_table;
+```
+
+In the SQL statement:
+
+- `TO MAXIMIZE|MINIMIZE ...` defines an expression string that describes the objective. 
+    - The syntax `MAXIMIZE|MINIMIZE` is used to specify the objective sense. 
+    - In the expression, `SUM` means sum the value across all rows like normal SQL statements.
+- `CONSTRAINT ...` expression strings that describe the constraints, can have multiple `CONSTRAINT` clause lines.
+- `WITH` attributes:
+    - variables: **required**, specify one column that stores the variable name. Using comma to separate if there are multiple variables.
+    - var_type: **required**, specify the variable type for each variable, format like "column:type,column:type...",  the type can be `Integers`, `NonNegativeIntegers`, `Reals` etc.
+- `USING`: **optional**, solver tool to use, default: glpk.
+- `INTO result_table`: set the result table name.
+
+After the SQL statement finishes execution, the result table `result_table` should look like:
+
+| product | result |
+| ------  | ------ |
+| soldier | 20     |
+| train   | 60     |
+
+### Two Variables
+
+For a more general example that have two variables (plants and markets, see the example described in https://en.wikipedia.org/wiki/AMPL for details), we want to minimize the cost of shipment between plants and markets, we have three tables looks like below:
```

Comment:
I don't know if people in the area of mathematical programming call such a problem having **two variables**?  If they do, let us follow their convention; otherwise, it looks to me that this problem has only one variable *amount* which is a two-dimensional matrix, with one index **plant** and another one **market**.